### PR TITLE
Add basic performance (load) tests 

### LINKED ISSUE #183: Create basic base load test
**What**

Currently our tests only focus on 1 or 2 ingresses. We have no understanding how the service acts under any type of load. Now that we have metrics and dashboards we need a test to start applying load and show and find bottlenecks around performance and scaling and to begin understanding the behaviour as the load increases. 
The goal of this isn't to push the service to its limits initially but rather find major issues and measure improvements as we make them. Examples of known limitations are:
- TLS certificates
- AWS route53 calls 

- [ ] Create a simple load test that will create a specified number of namespaces and Ingresses  against a targeted KCP workspace
- [ ] Add this test as something we can optionally run as part of the e2e tests
- [ ] Add this test to a workflow in GH so that we can trigger it as needed or on a schedule

### PR DESCRIPTION
**Pull request by mikenairn**
Closes #183 

### Description of Changes
Adds basic performance tests to appying varying load to an existing glbc instance (Ingress and DNSRecord). Ingress test runs a full e2e, creating a deployment, service and x number of Ingresses controlled by the "TEST_INGRESS_COUNT" env var. This test is expected to hit all parts of the glbc system, and in particular put load on the TLS issuer. DNSRecord test only creates dns record resources, and the number created is controlled by the "TEST_DNSRECORD_COUNT" env var. This test only hits
the dns record parts of the system, and will put load onto the AWS APIs.

```
make performance AWS_DNS_PUBLIC_ZONE_ID=xyz GLBC_DOMAIN=my.glbc.domain TEST_DNSRECORD_COUNT=1 TEST_INGRESS_COUNT=1
```

###### Terminal 1
```
(export $(cat ./config/deploy/local/controller-config.env.ci | xargs) && ./bin/kcp-glbc --kubeconfig config/deploy/local/kcp.kubeconfig)
```

###### Terminal 2
```
$ ./bin/kubectl-kcp workspace use root:default:kcp-glbc-user
$ (export $(cat ./config/deploy/local/controller-config.env.ci | xargs) && make performance TEST_DNSRECORD_COUNT=1 TEST_INGRESS_COUNT=1)
go build -o bin ./cmd/...
Performance Test Start: 1659515365172
KUBECONFIG="/home/mnairn/go/src/github.com/kuadrant/kcp-glbc/.kcp/admin.kubeconfig" \
AWS_DNS_PUBLIC_ZONE_ID="FAKE_ZONE_ID" \
TEST_DNSRECORD_COUNT="1" \
TEST_INGRESS_COUNT="1" \
go test -count=1 -timeout 60m -v ./e2e/performance -tags=performance
=== RUN   TestDNSRecord
=== PAUSE TestDNSRecord
=== RUN   TestIngress
=== PAUSE TestIngress
=== CONT  TestDNSRecord
=== CONT  TestIngress
    ingress_test.go:61: Creating 1 Ingresses
=== CONT  TestDNSRecord
    dns_record_test.go:78: Creating 1 DNSRecords
--- PASS: TestDNSRecord (0.08s)
--- PASS: TestIngress (13.28s)
PASS
ok      github.com/kuadrant/kcp-glbc/e2e/performance    13.306s
Performance Test End: 1659515380343
```

### Remaining Work 

The following changes still need to be completed:

- [ ] Refactor to allow code reuse between e2e and perf tests

### CODE REVIEW

File: e2e/performance/ingress_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		Apply(test.Ctx(), DeploymentConfiguration(namespace.Name, name), ApplyOptions)
+	test.Expect(err).NotTo(HaveOccurred())
+
+	// Create test Service
+	_, err = test.Client().Core().CoreV1().Services(namespace.Name).
+		Apply(test.Ctx(), ServiceConfiguration(namespace.Name, name, map[string]string{}), ApplyOptions)
+	test.Expect(err).NotTo(HaveOccurred())
+
+	// Create Ingresses
+	for i := 1; i <= ingressCount; i++ {
```
Comment (maleck13) [inline e2e/performance/ingress_test.go:74]: same here

**Review by maleck13 — COMMENTED** [review]
generally looks good suggestion around the creation logic

File: e2e/performance/dns_record_test.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	namespace := test.NewTestNamespace()
+	test.Expect(namespace).NotTo(BeNil())
+
+	dnsRecordCount := env.GetEnvInt(TestDNSRecordCount, DefaultTestDNSRecordCount)
+	test.Expect(dnsRecordCount > 0).To(BeTrue())
+	test.T().Log(fmt.Sprintf("Creating %d DNSRecords", dnsRecordCount))
+
+	// Create DNSRecords
+	for i := 1; i <= dnsRecordCount; i++ {
+		dnsRecord := createTestDNSRecord(test, namespace, glbcDomain)
```
[the remark this answers is not in the document]
Comment (mikenairn) [reply e2e/performance/dns_record_test.go:81]: I had wondered about this, but wasn't sure about the number of routines that might be getting created and if that would be an issue. Depends what you are trying to get out of the test as well, doing it in parallel will put more pressure on the API Server itself, but not really make much difference to the reconciling of Ingresses and DNSRecords, could be wrong though. Happy to try it and see anyway.

[the remark this answers is not in the document]
Comment (maleck13) [reply e2e/performance/dns_record_test.go:81]: I wouldn't worry about the number of routines tbh. Go is designed deal with huge numbers of these and they are designed to be very lightweight

**Pull request merged**